src/workqueue/workqueue.c

```diff
@@ -0,0 +1,849 @@
+		// normal pool uses finer-grained wq locking and atomic memory ops.
+		pthread_spin_lock(&workq->sl);
+		if (STAILQ_EMPTY(&workq->item_listhead)) {
+			os_atomic_or(&normal_pool.mask, wq_index_bit, relaxed);
+		}
+		STAILQ_INSERT_TAIL(&workq->item_listhead, wi, item_entry);
+		pthread_spin_unlock(&workq->sl);
+
+		if (unlikely(runnable_workers < target_max_runnable_workers)) {
+			_dispatch_work_added();
+		}
+	}
+
+	return 0;
+}
+
+
+static void
+_dispatch_workqueue_add_control_workitem(void *op_code)
+{
+	dispatch_workitem_t wi = _dispatch_alloc_workitem(NULL, op_code);
+	dispatch_workqueue_t workq = normal_pool.queues[0];
+	unsigned int wq_index_bit = 1; // highest priority queue
+
+	pthread_spin_lock(&workq->sl);
+	if (STAILQ_EMPTY(&workq->item_listhead)) {
+		os_atomic_or(&normal_pool.mask, wq_index_bit, relaxed);
+	}
+	STAILQ_INSERT_TAIL(&workq->item_listhead, wi, item_entry);
+	pthread_spin_unlock(&workq->sl);
+
+	if (unlikely(runnable_workers < target_max_runnable_workers)) {
+		_dispatch_work_added();
+	}
+}
+
+
+static void
+_dispatch_spawn_thread(void *(start_func)(void*))
+{
+	pthread_t tid;
+	pthread_attr_t attr;
+	int r;
+
+	r = pthread_attr_init(&attr);
+	(void)dispatch_assume_zero(r);
+
+	r = pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
+	(void)dispatch_assume_zero(r);
+
+	r = pthread_create(&tid, &attr, start_func, NULL);
+	(void)dispatch_assume_zero(r);
+
+	r = pthread_attr_destroy(&attr);
+	(void)dispatch_assume_zero(r);
+}
+
+
+static void*
+_dispatch_overcommit_worker_main(void *context DISPATCH_UNUSED)
+{
+	int r;
+	sigset_t mask;
+
+	r = sigfillset(&mask);
+	(void)dispatch_assume_zero(r);
+	r = _dispatch_pthread_sigmask(SIG_BLOCK, &mask, NULL);
+	(void)dispatch_assume_zero(r);
+
+	r = pthread_mutex_lock(&overcommit_pool.mutex);
+	(void)dispatch_assume_zero(r);
+
+	while (true) {
+		unsigned int idx = ffs(overcommit_pool.mask);
+		if (idx > 0) {
+			dispatch_workqueue_t wq = overcommit_pool.queues[idx - 1];
+			dispatch_workitem_t work = STAILQ_FIRST(&wq->item_listhead);
+			if (work != NULL) {
+				/* Remove the first work item */
+				STAILQ_REMOVE_HEAD(&wq->item_listhead, item_entry);
+				if (STAILQ_EMPTY(&wq->item_listhead)) {
+					overcommit_pool.mask &= ~(0x1 << wq->priority);
+				}
+
+				/* Release pool mutex */
+				r = pthread_mutex_unlock(&overcommit_pool.mutex);
+				(void)dispatch_assume_zero(r);
+
+				/* Execute the work item */
+				(void)dispatch_assume_zero(r);
+				void (*func)(void *) = work->func;
+				void *func_arg = work->func_arg;
+				_dispatch_dealloc_workitem(work);
+				func(func_arg);
+
+				/* Acquire pool mutex */
+				r = pthread_mutex_lock(&overcommit_pool.mutex);
+				(void)dispatch_assume_zero(r);
+				continue;
+			}
+		}
+
+		/* Wait for more work to be available. */
+		overcommit_pool.num_spares++;
+		r = pthread_cond_wait(&overcommit_pool.spare_workers, &overcommit_pool.mutex);
+	}
+}
+
+
+static void*
+_dispatch_normal_worker_main(void *context DISPATCH_UNUSED)
```

Comment:
I dislike having that much code duplicated, can't we do something where you have a generic `DISPATCH_ALWAYS_INLINE` function that takes a `bool overcommit` and handles the semantics differences inline, and then call this symbol from `_dispatch_normal_worker_main` and `_dispatch_overcommit_worker_main`? it will yield exactly the same code since the compiler will inline and constant fold the `overcommit` but it'll make maintenance much easier.

look at `_dispatch_queue_*_drain` for an example of that optimization technique.